### PR TITLE
Add Frame Toggling

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #2662

### Summary

* Add tests
* Add flow types
* Refactor Frames

### Test Plan

* unit tests for frame colllapsing

### CODE REVIEW

File: src/components/SecondaryPanes/Frames.js [hunk 1]
```diff
@@ -1,239 +0,0 @@
-// @flow
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/Frames.js]: Frames was moved to index and Frame was pulled out

File: src/components/SecondaryPanes/Frames/index.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  return frames;
+}
+
+export default connect(
+  state => ({
+    frames: getAndProcessFrames(state),
+    selectedFrame: getSelectedFrame(state)
+  }),
+  dispatch => bindActionCreators(actions, dispatch)
+)(Frames);
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/Frames/index.js]: Smaller and focused on handling frame lists

File: src/components/SecondaryPanes/Frames/tests/Frames.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        selectedFrame,
+        contextTypes: {},
+        selectFrame: jest.fn()
+      })
+    );
+    expect(component).toMatchSnapshot();
+    // const showMore = component.node.props.children[1];
+    // expect(showMore.props.className).toEqual("show-more");
+  });
+});
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/Frames/tests/Frames.js]: lots of tests :)

File: src/components/SecondaryPanes/WhyPaused.js [hunk 4]
```diff
@@ -23,7 +24,7 @@ function renderExceptionSummary(exception) {
 }
 
 class WhyPaused extends Component {
-  renderMessage(pauseInfo) {
+  renderMessage(pauseInfo: Pause) {
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/WhyPaused.js]: just adding more flow coverage

File: src/test/tests-setup.js [hunk 5]
```diff
@@ -33,15 +33,16 @@ const { setConfig } = require("devtools-config");
...
+const config = Object.assign({}, envConfig, {
+  workers: {
+    sourceMapURL: path.join(
+      rootPath,
+      "node_modules/devtools-source-map/src/worker.js"
+    ),
+    parserURL: path.join(rootPath, "src/utils/parser/worker.js"),
+    prettyPrintURL: path.join(rootPath, "src/utils/pretty-print/worker.js")
+  }
+});
```
Comment (jasonLaster) [inline src/test/tests-setup.js]: needed this because Frames depends on launchpad which wants DebuggerConfig set

File: src/utils/frame.js [hunk 6]
```diff
@@ -88,12 +112,64 @@ export function simplifyDisplayName(displayName) {
...
-  if (library) {
+  if (library && shouldMapDisplayName) {
     displayName = mapDisplayNames(frame, library);
   }
 
   displayName = simplifyDisplayName(displayName);
   return endTruncateStr(displayName, 25);
 }
+
+export function collapseFrames(frames) {
```
Comment (jasonLaster) [inline src/utils/frame.js]: yes this is ugly plane code, but it passes tests and we can fix it up another time :)

unless you can help me do better

File: src/components/SecondaryPanes/Frames/Frame.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        label: copySourceUrlLabel,
+        accesskey: copySourceUrlKey,
+        disabled: false,
+        click: () => copyToTheClipboard(source.url)
+      };
+
+      menuOptions.push(copySourceUrl);
+    }
+
+    const framesToCopy = frames.map(f => formatCopyName(f)).join("\n");
```
Comment (bomsy) [inline src/components/SecondaryPanes/Frames/Frame.js]: looks like `frames` does not get passed into this `onContextMenu` function

**Comment by bomsy** [comment]
Frame collapsing is cool!!

**Comment by bomsy** [comment]
hmmm .... the tests seem to fail for me.

**Pull request merged**